File: app/src/test/java/com/kira/kdownloader/engine/DownloadOutputSelectorTest.java
package com.kira.kdownloader.engine;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class DownloadOutputSelectorTest {
    @Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();

    @Test public void videoSelectionIgnoresAudioAndIntermediateDashStreams() throws IOException {
        mediaFile("reel.f123.mp4", 300L);
        mediaFile("reel.m4a", 400L);
        File completed = mediaFile("reel.mp4", 100L);
        assertEquals(completed, DownloadOutputSelector.select(Arrays.asList(temporaryFolder.getRoot().listFiles()), FormatSelector.Kind.VIDEO));
    }

    @Test public void videoSelectionFailsInsteadOfPublishingAnAudioOnlyResult() throws IOException {
        mediaFile("reel.m4a", 100L);
        mediaFile("reel.f123.mp4", 100L);
        assertNull(DownloadOutputSelector.select(Arrays.asList(temporaryFolder.getRoot().listFiles()), FormatSelector.Kind.VIDEO));
    }

    @Test public void audioSelectionPrefersTheConvertedMp3() throws IOException {
        mediaFile("track.m4a", 500L);
        File mp3 = mediaFile("track.mp3", 100L);
        assertEquals(mp3, DownloadOutputSelector.select(Arrays.asList(temporaryFolder.getRoot().listFiles()), FormatSelector.Kind.AUDIO));
    }

    private File mediaFile(String name, long modifiedAt) throws IOException {
        File file = temporaryFolder.newFile(name);
        Files.write(file.toPath(), "media".getBytes(StandardCharsets.UTF_8));
        file.setLastModified(modifiedAt);
        return file;
    }
}

File: app/src/test/java/com/kira/kdownloader/engine/DownloadOutputSelectorTest.kt
package com.kira.kdownloader.engine

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import java.io.File

class DownloadOutputSelectorTest {
    @get:Rule
    val temporaryFolder = TemporaryFolder()

    @Test
    fun `video selection ignores audio and intermediate DASH streams`() {
        mediaFile("reel.f123.mp4", modifiedAt = 300L)
        mediaFile("reel.m4a", modifiedAt = 400L)
        val completedVideo = mediaFile("reel.mp4", modifiedAt = 100L)

        assertEquals(
            completedVideo,
            DownloadOutputSelector.select(
                temporaryFolder.root.listFiles().orEmpty().toList(),
                FormatSelector.Kind.VIDEO,
            ),
        )
    }

    @Test
    fun `video selection fails instead of publishing an audio-only result`() {
        mediaFile("reel.m4a")
        mediaFile("reel.f123.mp4")

        assertNull(
            DownloadOutputSelector.select(
                temporaryFolder.root.listFiles().orEmpty().toList(),
                FormatSelector.Kind.VIDEO,
            ),
        )
    }

    @Test
    fun `audio selection prefers the converted mp3`() {
        mediaFile("track.m4a", modifiedAt = 500L)
        val mp3 = mediaFile("track.mp3", modifiedAt = 100L)

        assertEquals(
            mp3,
            DownloadOutputSelector.select(
                temporaryFolder.root.listFiles().orEmpty().toList(),
                FormatSelector.Kind.AUDIO,
            ),
        )
    }

    private fun mediaFile(name: String, modifiedAt: Long = 100L): File =
        temporaryFolder.newFile(name).apply {
            writeText("media")
            setLastModified(modifiedAt)
        }
}
